packages/apollo-engine-reporting/src/extension.ts

```diff
@@ -199,12 +199,18 @@ export class EngineReportingExtension<TContext = any>
       this.trace.fullQueryCacheHit = !!o.requestContext.metrics
         .responseCacheHit;
 
-      const operationName = this.operationName || '';
+      // If the `operationName` was not already set elsewhere, for example,
+      // through the `executionDidStart` or the `willResolveField` hooks, then
+      // we'll resort to using the `operationName` which was requested to be
```

Comment:
I'm not sure this change does what the comment says. `requestContext.operationName` is not "the name the client wrote in the operation" (that's `request.operationName`). It's "the name of the operation that will be run, assuming the request passes the parsing and validation stages, and the 'operation naming' rules pass".

By "operation naming rules" I mean "either the document contains an operation with the given name, or no name was given and there is exactly one operation".  Notably, this is null if the client wrote a name but the name isn't present.

`this.operationName` currently gets set to:
  - if we make it to executionDidStart, "the name the client wrote in the operation" (because that's how ExecutionArgs is set up in requestPipeline: request.operationName, not requestContext.operationName)
  - if the client didn't write a name in the operation and we make it to resolve some field, it gets set to the actual operation name we're running (which, in the case we made it this far, must equal requestContext.operationName)

Notably, though, parsing or validation fails, then not only will we fail to make it to executionDidStart/willResolveField, we also won't get to the code that assigns requestContext.operationName. And so the code here won't actually add the client-supplied operation name, which I think is the goal? Or is it only that we care about the didResolveOperation throwing case, and not the parse failure / validation failure case?  (THANK YOU for explaining in the PR desc the motivation behind this PR!!!)

Anyway, I'm trying right now to disengage willResolveField (timing calculations) from the rest of the extension (a refactoring that'll be helpful for federated metrics).  I *think* I have a simple answer and will send you a PR including a tweak of this comment. See also https://github.com/apollographql/apollo-server/pull/2557#issuecomment-504583407

Reply:
Thanks for sorting this out, @glasser!